File: flutter_zoom_videosdk-2.4.0/ios/Classes/FlutterZoomVideoSdkSubSessionHelper.m
#import "FlutterZoomVideoSdkSubSessionHelper.h"
#import "FlutterZoomVideoSdkSubSessionKit.h"
#import "FlutterZoomVideoSdkSubSessionUser.h"
#import "JSONConvert.h"

@implementation FlutterZoomVideoSdkSubSessionHelper {
    NSMutableDictionary<NSString *, ZoomVideoSDKSubSessionKit *> *subSessionKitMap;
}

static ZoomVideoSDKSubSessionManager *subSessionManager;
static ZoomVideoSDKSubSessionParticipant *subSessionParticipant;
static ZoomVideoSDKSubSessionUserHelpRequestHandler *subSessionUserHelpRequestHandler;

- (instancetype)init {
    self = [super init];
    if (self) {
        subSessionKitMap = [[NSMutableDictionary alloc] init];
    }
    return self;
}

- (ZoomVideoSDKSubSessionHelper *)getSubSessionHelper {
    ZoomVideoSDKSubSessionHelper *subSessionHelper = nil;

    @try {
        subSessionHelper = [[ZoomVideoSDK shareInstance] getsubSessionHelper];
        if (subSessionHelper == nil) {
            @throw [NSException exceptionWithName:@"SubSessionHelperException"
                                           reason:@"No SubSession Helper Found"
                                         userInfo:nil];
        }
    } @catch (NSException *exception) {
        NSLog(@"Exception: %@", exception);
    }

    return subSessionHelper;
}

+(void) storeSubSessionUserHelpRequestHandler:(ZoomVideoSDKSubSessionUserHelpRequestHandler *)handler {
    subSessionUserHelpRequestHandler = handler;
}

+(void) storeSubSessionManager:(ZoomVideoSDKSubSessionManager *)manager {
    subSessionManager = manager;
}

+(void) storeSubSessionParticipant:(ZoomVideoSDKSubSessionParticipant *)participant {
    subSessionParticipant = participant;
}

-(void) joinSubSession:(FlutterMethodCall *)call withResult:(FlutterResult)result {
    NSDictionary *args = call.arguments;
    if (args == nil) {
        result([FlutterError errorWithCode:[[JSONConvert ZoomVideoSDKErrorValuesReversed] objectForKey:@(Errors_Invalid_Parameter)]
                                   message:@"No call arguments"
                                   details:nil]);
        return;
    }

    NSString *subSessionId = args[@"subSessionId"];
    ZoomVideoSDKSubSessionKit *subSessionKit = subSessionKitMap[subSessionId];
    if (subSessionKit == nil) {
        result([FlutterError errorWithCode:[[JSONConvert ZoomVideoSDKErrorValuesReversed] objectForKey:@(Errors_Invalid_Parameter)]
                                   message:@"No SubSessionKit found for the given subSessionId"
                                   details:nil]);
        return;
    }

    dispatch_async(dispatch_get_main_queue(), ^{
        result([[JSONConvert ZoomVideoSDKErrorValuesReversed] objectForKey:@([subSessionKit joinSubSession])]);
    });
}

-(void) startSubSession:(FlutterResult)result {
    dispatch_async(dispatch_get_main_queue(), ^{
        result([[JSONConvert ZoomVideoSDKErrorValuesReversed] objectForKey:@([subSessionManager startSubSession])]);
    });
}

-(void) isSubSessionStarted:(FlutterResult)result {
    if ([subSessionManager isSubSessionStarted]) {
        result(@YES);
    } else {
        result(@NO);
    }
}

-(void) stopSubSession:(FlutterResult)result {
    dispatch_async(dispatch_get_main_queue(), ^{
        result([[JSONConvert ZoomVideoSDKErrorValuesReversed] objectForKey:@([subSessionManager stopSubSession])]);
    });
}

-(void) broadcastMessage:(FlutterMethodCall *)call withResult:(FlutterResult)result {
    NSDictionary *args = call.arguments;
    NSString *message = args[@"message"];

    dispatch_async(dispatch_get_main_queue(), ^{
        result([[JSONConvert ZoomVideoSDKErrorValuesReversed] objectForKey:@([subSessionManager broadcastMessage:message])]);
    });
}

-(void) returnToMainSession:(FlutterResult)result {
    dispatch_async(dispatch_get_main_queue(), ^{
        result([[JSONConvert ZoomVideoSDKErrorValuesReversed] objectForKey:@([subSessionParticipant returnToMainSession])]);
    });
}

-(void) requestForHelp:(FlutterResult)result {
    dispatch_async(dispatch_get_main_queue(), ^{
        result([[JSONConvert ZoomVideoSDKErrorValuesReversed] objectForKey:@([subSessionParticipant requestForHelp])]);
    });
}

-(void) getRequestUserName:(FlutterResult)result {
    result([subSessionUserHelpRequestHandler getRequestUserName]);
}

-(void) getRequestSubSessionName:(FlutterResult)result {
    result([subSessionUserHelpRequestHandler getRequestSubSessionName]);
}

-(void) ignoreUserHelpRequest:(FlutterResult)result {
    dispatch_async(dispatch_get_main_queue(), ^{
        result([[JSONConvert ZoomVideoSDKErrorValuesReversed] objectForKey:@([subSessionUserHelpRequestHandler ignore])]);
    });
}

-(void) joinSubSessionByUserRequest:(FlutterResult)result {
    dispatch_async(dispatch_get_main_queue(), ^{
        result([[JSONConvert ZoomVideoSDKErrorValuesReversed] objectForKey:@([subSessionUserHelpRequestHandler joinSubSessionByUserRequest])]);
    });
}

-(void) commitSubSessionList:(FlutterMethodCall *)call withResult:(FlutterResult)result {
    NSArray<NSString *> *subSessionNames = call.arguments;

    dispatch_async(dispatch_get_main_queue(), ^{
        ZoomVideoSDKError err = [[self getSubSessionHelper] addSubSessionToPreList: subSessionNames];
        if (err == Errors_Success) {
            result([[JSONConvert ZoomVideoSDKErrorValuesReversed] objectForKey:@([[self getSubSessionHelper] commitSubSessionList])]);
        } else {
            result([FlutterError errorWithCode:[[JSONConvert ZoomVideoSDKErrorValuesReversed] objectForKey:@(err)]
                                       message:@"Not able to commit SubSessionList"
                                       details:nil]);
        }
        
    });
}

-(void) getCommittedSubSessionList:(FlutterResult)result {
    dispatch_async(dispatch_get_main_queue(), ^{
        NSArray<ZoomVideoSDKSubSessionKit *> *subSessionKits = [[self getSubSessionHelper] getCommittedSubSessionList];
        for (ZoomVideoSDKSubSessionKit *kit in subSessionKits) {
            [self->subSessionKitMap setObject:kit forKey:[kit getSubSessionID]];
        }
        result([FlutterZoomVideoSdkSubSessionKit mapSubSessionKitArray:subSessionKits]);
    });
}

-(void) withdrawSubSessionList:(FlutterResult)result {
    dispatch_async(dispatch_get_main_queue(), ^{
        result([[JSONConvert ZoomVideoSDKErrorValuesReversed] objectForKey:@([[self getSubSessionHelper] withdrawSubSessionList])]);
    });
}

@end

File: flutter_zoom_videosdk-2.4.0/android/src/main/java/com/flutterzoom/videosdk/FlutterZoomVideoSdkSubSessionHelper.java
package com.flutterzoom.videosdk;

import android.app.Activity;

import androidx.annotation.NonNull;

import com.flutterzoom.videosdk.convert.FlutterZoomVideoSdkErrors;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import us.zoom.sdk.SubSessionKit;
import us.zoom.sdk.SubSessionUserHelpRequestHandler;
import us.zoom.sdk.ZoomVideoSDK;
import us.zoom.sdk.ZoomVideoSDKErrors;
import us.zoom.sdk.ZoomVideoSDKSubSessionHelper;
import us.zoom.sdk.ZoomVideoSDKSubSessionManager;
import us.zoom.sdk.ZoomVideoSDKSubSessionParticipant;

public class FlutterZoomVideoSdkSubSessionHelper {

    private Activity activity;
    private static ZoomVideoSDKSubSessionManager subSessionManager;
    private static ZoomVideoSDKSubSessionParticipant subSessionParticipant;
    private static SubSessionUserHelpRequestHandler subSessionUserHelpRequestHandler;
    private Map<String, SubSessionKit> subSessionKitMap;

    FlutterZoomVideoSdkSubSessionHelper(Activity activity) {
        this.activity = activity;
        subSessionKitMap = new HashMap<>();
    }

    private ZoomVideoSDKSubSessionHelper getSubSessionHelper() {
        ZoomVideoSDKSubSessionHelper subSessionHelper = null;

        try {
            subSessionHelper = ZoomVideoSDK.getInstance().getSubSessionHelper();
            if (subSessionHelper  == null) {
                throw new Exception("No SubSession Helper Found");
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return subSessionHelper;
    }

    public static void storeSubSessionUserHelpRequestHandler(SubSessionUserHelpRequestHandler handler) {
        subSessionUserHelpRequestHandler = handler;
    }

    public static void storeSubSessionManager(ZoomVideoSDKSubSessionManager manager) {
        subSessionManager = manager;
    }

    public static void storeSubSessionParticipant(ZoomVideoSDKSubSessionParticipant participant) {
        subSessionParticipant = participant;
    }

    public void joinSubSession(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
        Map<String, Object> args = call.arguments();
        if (args == null) {
            result.error(FlutterZoomVideoSdkErrors.valueOf(ZoomVideoSDKErrors.Errors_Invalid_Parameter), "No call arguments", null);
            return;
        }
        String subSessionId = (String) args.get("subSessionId");
        SubSessionKit subSessionKit = subSessionKitMap.get(subSessionId);
        if (subSessionKit == null) {
            result.error(FlutterZoomVideoSdkErrors.valueOf(ZoomVideoSDKErrors.Errors_Invalid_Parameter), "No SubSessionKit found for the given subSessionId", null);
            return;
        }

        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                result.success(FlutterZoomVideoSdkErrors.valueOf(subSessionKit.joinSubSession()));
            }
        });
    }

    public void startSubSession(@NonNull MethodChannel.Result result) {
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                result.success(FlutterZoomVideoSdkErrors.valueOf(subSessionManager.startSubSession()));
            }
        });
    }

    public void isSubSessionStarted(@NonNull MethodChannel.Result result) {
        result.success(subSessionManager.isSubSessionStarted());
    }

    public void stopSubSession(@NonNull MethodChannel.Result result) {
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                result.success(FlutterZoomVideoSdkErrors.valueOf(subSessionManager.stopSubSession()));
            }
        });
    }

    public void broadcastMessage(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
        Map<String, Object> args = call.arguments();
        String message = (String) args.get("message");

        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                result.success(FlutterZoomVideoSdkErrors.valueOf(subSessionManager.broadcastMessage(message)));
            }
        });
    }

    public void returnToMainSession(@NonNull MethodChannel.Result result) {
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                result.success(FlutterZoomVideoSdkErrors.valueOf(subSessionParticipant.returnToMainSession()));
            }
        });
    }

    public void requestForHelp(@NonNull MethodChannel.Result result) {
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                result.success(FlutterZoomVideoSdkErrors.valueOf(subSessionParticipant.requestForHelp()));
            }
        });
    }

    public void getRequestUserName(@NonNull MethodChannel.Result result) {
        result.success(subSessionUserHelpRequestHandler.getRequestUserName());
    }

    public void getRequestSubSessionName(@NonNull MethodChannel.Result result) {
        result.success(subSessionUserHelpRequestHandler.getRequestSubSessionName());
    }

    public void ignoreUserHelpRequest(@NonNull MethodChannel.Result result) {
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                result.success(FlutterZoomVideoSdkErrors.valueOf(subSessionUserHelpRequestHandler.ignore()));
            }
        });
    }

    public void joinSubSessionByUserRequest(@NonNull MethodChannel.Result result) {
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                result.success(FlutterZoomVideoSdkErrors.valueOf(subSessionUserHelpRequestHandler.joinSubSessionByUserRequest()));
            }
        });
    }

    public void commitSubSessionList(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
        ArrayList<String> subSessionNames = (ArrayList<String>) call.arguments;

        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                result.success(FlutterZoomVideoSdkErrors.valueOf(getSubSessionHelper().commitSubSessionList(subSessionNames)));
            }
        });
    }

    public void getCommittedSubSessionList(@NonNull MethodChannel.Result result) {
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                List<SubSessionKit> subSessionKits = getSubSessionHelper().getCommittedSubSessionList();
                subSessionKits.forEach(kit -> subSessionKitMap.put(kit.getSubSessionID(), kit));
                result.success(FlutterZoomVideoSdkSubSessionKit.jsonSubSessionKitArray(subSessionKits));
            }
        });
    }

    public void withdrawSubSessionList(@NonNull MethodChannel.Result result) {
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                result.success(FlutterZoomVideoSdkErrors.valueOf(getSubSessionHelper().withdrawSubSessionList()));
            }
        });
    }

}
